### PR TITLE
Use MD5 to match image file names with exported csv records

### PR DESCRIPTION
**Pull request by step-**
This PR is work in progress.  I will complete it if you think it's worthy and if you give me directions.  See this PR's discussion for details.

Is it worth doing?  I think so.  This is the problem I need to solve: matching exported image file names with their corresponding exported csv record.

**My analysis of the master version** -- The image file name consists of the name+surname, which isn't unique, and a random unique id, which can't be matched with any csv record. The latter happens because 1) records don't include such ids, and 2) the unique id is randomly generated when the image is exported, which happens separately from csv export.  Feel free to correct my analysis.

So, in order to match an image with its record, we need an image id that can be generated consistently when image exporting and csv exporting take place separately.  Alternatively, we need to export the csv and the img formats in a single run, and save the current uniqid in the csv record.  My preference is to get rid of the uniqid altogether, because it's random, and replace it with md5, because it can be consistently generated in the same run or even in separate runs.

**What's missing in the current PR** --  Apart from deleting lines 933-936, I need to know from you how to add a new csv column that holds the md5 of card->photo data.  Moreover, what should the new csv column's name?  Finally,  I'm not familiar with the other formats, ldif, fritzbox, etc. so I don't know if an md5 should or even could be added to such export formats.

In conclusion, are you interested in this feature, would you do it with an md5, and can you tell me where/what to change to add the new csv column?  I won't be offended if you prefer to make the changes yourself :)

### CODE REVIEW

**Comment by thomascube** [comment]
Hi @step- 

You're PR and the use-case behind surely makes sense an the use of a unique and non-random identifier for a photo (or better a vcard) is a requirement for it. Rather than computing the MD5 sum of the photo property, using or generating a unique identifier for the vcard record itself seems a better approach to me which is also generally useful in a CSV export. The VCard format actually defines a [UID property](https://tools.ietf.org/html/rfc6350#section-6.7.6) but that's not a mandatory field and might be missing. If present, that value should used and if absent, generated as a checksum (e.g. md5) over all (or some selected) properties of the vcard.

For a proper solution I suggest the following:
- check if a UID is present and if not, generate one from the vcard properties (so that the same vcard receives the same uid on every conversion run). The LDIF export requires a uid and the converter [currently generates a random id](https://github.com/thomascube/vcfconvert/blob/master/vcard_convert.php#L699). This should be moved to the `normalize()` function and refactored accordingly.
- use the `$card->uid` for the image filename instead of the random id. I'd keep the name part of the image filename for those who just want to pick a photo by name from the export.
- add a *uid* column on the `toCSV()` function by simply adding another value to the header line and each card line (append to `$out`.

Many thanks for your inputs and your contribution!

**Comment by step-** [comment]
Hi Thomas, thank you for your comments.  Your plan is good.  I need to add two things.

The unique id we generate will consist of  two parts separated by `-`.  As you said, part 1 will be an MD5 of some properties, to be defined.  Special case: if a card does have a phone in it, the MD5 will be of the photo data itself.  Reason: a script will be able to detect duplicate pictures by just looking at part 1 of the unique id included in the image file name.  Part 2 of the unique id will still be generated randomly.  Reason: avoid collisions when two vcards contain exactly the same data - we can't assume they won't.

The unique id will be stored in a new vcard property, namely "vcfconvert_uid" for lack of a better name.   We will not reuse the existing uid name.  Reasons: because uid isn't ours.  We can't overwrite it with our id if it isn't empty.  Also, uid's value doesn't have the two-part structure outlines in the previous paragraph, so it doesn't enable detecting duplicate photos by looking at their part 1 file name element.

If you agree on both paragraphs above I can start implementing, and I will ask more questions if I get stuck.  Thank you.

**Comment by step-** [comment]
So I though some more about the point I left open in my previous comment.  You [wrote](https://github.com/thomascube/vcfconvert/pull/19#issuecomment-739551794):

> Rather than computing the MD5 sum of the photo property, using or generating a unique identifier for the vcard record itself seems a better approach to me which is also generally useful in a CSV export.

Useful for what purpose?  It seems to me that generating a general `vcfconvert_uid` for all records during export serves no useful purpose.  The CVS is not the source of the data but just and export of the data.  That the VCard data source _sometimes_ includes a `uid` value is not our concern. We just export values.  We could _use_ the VCard's uid for our own needs the uid meets our needs.  However, the VCard uid doesn't meet the need expressed in my [original proposal](https://github.com/thomascube/vcfconvert/pull/19#issue-533175689).  For vcfconvert to add a `vcfconvert_uid` we must have a purpose otherwise we better not.  I have no purpose for it and I can't see one.

My purpose for adding an MD5 is identifying the exported photo while linking such "photo_data_md5", exposed in the image file name, with the csv record from which the image file originates.

> The VCard format actually defines a UID property but that's not a mandatory field and might be missing. If present, that value should used and if absent, generated as a checksum (e.g. md5) over all (or some selected) properties of the vcard.

Again, I don't see a reason to fill the VCard `uid` with our own arbitrarily generated value for all records, and we can't reuse the VCard `uid` value because there is no guarantee that the value will be available.

> For a proper solution I suggest the following:
>
> * check if a UID is present and if not, generate one from the vcard properties (so that the same vcard receives the same uid on every conversion run). The LDIF export requires a uid and the converter currently generates a random id. This should be moved to the normalize() function and refactored accordingly.
> * use the $card->uid for the image filename instead of the random id. I'd keep the name part of the image filename for those who just want to pick a photo by name from the export.
> * add a uid column on the toCSV() function by simply adding another value to the header line and each card line

I counter-propose the following:
>
> * ~~check if a UID is present and if not, generate one from the vcard properties (so that the same vcard receives the same uid on every conversion run).~~ The LDIF export requires a uid and the converter currently generates a random id. ~~This should be moved to the normalize() function and refactored accordingly.~~
> * ~~use the $card->uid for the image filename instead of the random id.~~ I'd keep the name part of the image filename for those who just want to pick a photo by name from the export.
> * _use `md5($card->photo->...data)` for the image filename instead of the random id keeping the name part of the image filename._
> * add a ~~uid~~ `photo_data_md5` column on the toCSV() function by simply adding another value to the header line and each card line

I like the idea of dropping the `uniqid()` from the image filename because this implies that we create just one image files for all records that export the same photo.

**Comment by thomascube** [comment]
My suggestions want to keep a general purpose in mind while providing a solution to your - rather specific - needs to exporting into CSV and photo images. Fee free to implement it your way but the more generally useful the changes, the higher is the likeliness of your PR being accepted upstream.

**Comment by step-** [comment]
> My suggestions want to keep a general purpose in mind
Yes, I understand.   Generality with a purpose can be a good thing.  It's only that I can't think of a practical application where generating a generic `vcfconvert_uid` would be useful.  How do you envision it being used?

**Comment by thomascube** [comment]
> How do you envision it being used?

Having an (unique) identifier for a data record can be useful in many ways. For example when upadating a data set. LDAP for example requires one and the converter currently fails to provide a stable identifier unless already present in the vcard.

You're apparently only looking for an identifier for the photo source which is a slightly different use-case but in my opinion could be solved with the UID concept I have in mind. Anyway, you may choose your way and in that case I suggest to name the column `photo_id` in the CSV output.

**Comment by step-** [comment]
Well in fact I think I will add a new column named `photo_filename` and leave identity concerns to future developers who have a need for such an id.  For `photo_filename` to show consistent values across different runs, the file name will consist of `display name` + `md5` without random components added to the photo file name.  I will follow up on this PR in a few days as I'm quite busy now.  Thank you.
